src/meta-srv/src/region/supervisor.rs

```diff
@@ -0,0 +1,575 @@
+        kv_backend: KvBackendRef,
+    ) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::channel(1024);
+        Self {
+            failure_detector: RegionFailureDetector::new(options),
+            tick_interval,
+            receiver: rx,
+            sender: tx,
+            selector_context,
+            selector,
+            region_migration_manager,
+            table_metadata_manager: Arc::new(TableMetadataManager::new(kv_backend.clone())),
+            kv_backend,
+        }
+    }
+
+    /// Returns the [`HeartbeatAcceptor`].
+    pub(crate) fn heartbeat_acceptor(&self) -> HeartbeatAcceptor {
+        HeartbeatAcceptor {
+            sender: self.sender.clone(),
+        }
+    }
+
+    /// Returns the [`RegionSupervisorTicker`].
+    pub(crate) fn ticker(&self) -> RegionSupervisorTickerRef {
+        Arc::new(RegionSupervisorTicker {
+            tick_interval: self.tick_interval,
+            sender: self.sender.clone(),
+            tick_handle: Mutex::new(None),
+        })
+    }
+
+    /// Runs the main loop.
+    pub(crate) async fn run(&mut self) {
+        while let Some(event) = self.receiver.recv().await {
+            match event {
+                Event::Tick => {
+                    let regions = self.detect_region_failure();
+                    self.handle_region_failures(regions).await;
+                }
+                Event::HeartbeatArrived(heartbeat) => self.on_heartbeat_arrived(heartbeat),
+                Event::Clear => self.clear(),
+                #[cfg(test)]
+                Event::Dump(sender) => {
+                    let _ = sender.send(self.failure_detector.dump());
+                }
+            }
+        }
+        info!("RegionSupervisor is stopped!");
+    }
+
+    async fn handle_region_failures(&self, regions: Vec<(ClusterId, DatanodeId, RegionId)>) {
+        if regions.is_empty() {
+            return;
+        }
+        match self.is_maintenance_mode().await {
+            Ok(false) => {}
+            Ok(true) => {
+                info!("Maintenance mode is enabled, skip failover");
+                return;
+            }
+            Err(err) => {
+                error!(err; "Failed to check maintenance mode");
+                return;
+            }
+        }
+
+        let table_ids = regions
+            .iter()
+            .map(|(_, _, region_id)| region_id.table_id())
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        let table_routes = match self
+            .table_metadata_manager
+            .table_route_manager()
+            .batch_get_physical_table_routes(&table_ids)
+            .await
+            .context(error::TableMetadataManagerSnafu)
+        {
+            Ok(table_routes) => table_routes,
+            Err(err) => {
+                error!(err; "Failed to retrieves table routes: {table_ids:?}");
+                return;
+            }
+        };
```

Comment:
It's better to retrieve the table_routes in the `submit_produce` method and handle this in the background.